File: hw/src/main/java/com/sumian/app/base/BaseActivity.java
package com.sumian.app.base;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Looper;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v7.app.AppCompatActivity;
import android.view.Gravity;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;

import com.sumian.app.common.helper.ToastHelper;
import com.sumian.app.log.LogJobIntentService;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * Created by jzz
 * on 2017/4/left.
 * <p>
 * desc:
 */

public abstract class BaseActivity<Presenter> extends AppCompatActivity {

    private static final String TAG = BaseActivity.class.getName();
    private Unbinder mBind;
    private View mRoot;

    protected Presenter mPresenter;

    /**
     * 设置状态栏黑色字体图标，
     * 适配4.4以上版本MIUIV、Flyme和6.0以上版本其他Android
     *
     * @param activity activity
     * @return 1:MIUUI 2:Flyme 3:android6.0
     */
    public static int StatusBarLightMode(Activity activity) {
        int result = 0;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            if (MIUISetStatusBarLightMode(activity.getWindow(), true)) {
                result = 1;
            } else if (FlymeSetStatusBarLightMode(activity.getWindow(), true)) {
                result = 2;
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                activity.getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR);
                result = 3;
            }
        }
        return result;
    }

    /**
     * 已知系统类型时，设置状态栏黑色字体图标。
     * 适配4.4以上版本MIUIV、Flyme和6.0以上版本其他Android
     *
     * @param activity activity
     * @param type     1:MIUUI 2:Flyme 3:android6.0
     */
    public static void StatusBarLightMode(Activity activity, int type) {
        if (type == 1) {
            MIUISetStatusBarLightMode(activity.getWindow(), true);
        } else if (type == 2) {
            FlymeSetStatusBarLightMode(activity.getWindow(), true);
        } else if (type == 3) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                activity.getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR);
            }
        }

    }

    /**
     * 清除MIUI或flyme或6.0以上版本状态栏黑色字体
     */
    public static void StatusBarDarkMode(Activity activity, int type) {
        if (type == 1) {
            MIUISetStatusBarLightMode(activity.getWindow(), false);
        } else if (type == 2) {
            FlymeSetStatusBarLightMode(activity.getWindow(), false);
        } else if (type == 3) {
            activity.getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_VISIBLE);
        }

    }

    /**
     * 设置状态栏字体图标为深色，需要MIUIV6以上
     *
     * @param window 需要设置的窗口
     * @param dark   是否把状态栏字体及图标颜色设置为深色
     * @return boolean 成功执行返回true
     */
    @SuppressWarnings("unchecked")
    public static boolean MIUISetStatusBarLightMode(Window window, boolean dark) {
        boolean result = false;
        if (window != null) {
            Class clazz = window.getClass();
            try {
                int darkModeFlag;
                @SuppressLint("PrivateApi") Class layoutParams = Class.forName("android.view.MiuiWindowManager$LayoutParams");
                Field field = layoutParams.getField("EXTRA_FLAG_STATUS_BAR_DARK_MODE");
                darkModeFlag = field.getInt(layoutParams);
                Method extraFlagField = clazz.getMethod("setExtraFlags", int.class, int.class);
                if (dark) {
                    extraFlagField.invoke(window, darkModeFlag, darkModeFlag);//状态栏透明且黑色字体
                } else {
                    extraFlagField.invoke(window, 0, darkModeFlag);//清除黑色字体
                }
                result = true;
            } catch (Exception ignored) {

            }
        }
        return result;
    }

    /**
     * 设置状态栏图标为深色和魅族特定的文字风格，Flyme4.0以上
     * 可以用来判断是否为Flyme用户
     *
     * @param window 需要设置的窗口
     * @param dark   是否把状态栏字体及图标颜色设置为深色
     * @return boolean 成功执行返回true
     */
    public static boolean FlymeSetStatusBarLightMode(Window window, boolean dark) {
        boolean result = false;
        if (window != null) {
            try {
                WindowManager.LayoutParams lp = window.getAttributes();
                Field darkFlag = WindowManager.LayoutParams.class
                    .getDeclaredField("MEIZU_FLAG_DARK_STATUS_BAR_ICON");
                Field meizuFlags = WindowManager.LayoutParams.class
                    .getDeclaredField("meizuFlags");
                darkFlag.setAccessible(true);
                meizuFlags.setAccessible(true);
                int bit = darkFlag.getInt(null);
                int value = meizuFlags.getInt(lp);
                if (dark) {
                    value |= bit;
                } else {
                    value &= ~bit;
                }
                meizuFlags.setInt(lp, value);
                window.setAttributes(lp);
                result = true;
            } catch (Exception ignored) {

            }
        }
        return result;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        StatusBarDarkMode(this, StatusBarLightMode(this));
        if (initBundle(getIntent().getExtras())) {
            setContentView(getLayoutId());
            initWindow();
            this.mBind = ButterKnife.bind(this);
            this.mRoot = getWindow().getDecorView();
            initWidget();
            initPresenter();
            initData();
        } else {
            finish();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        boolean appIsInBackground = LogJobIntentService.isAppIsInBackground(this);
        if (appIsInBackground) {
            LogJobIntentService.enqueueWork(this, new Intent(this, LogJobIntentService.class));
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        onRelease();
        // this.mBind.unbind();
        //this.mRoot = null;
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        if (!initBundle(intent.getExtras())) {
            finish();
        }
    }

    protected void initWindow() {

    }

    protected abstract int getLayoutId();

    protected boolean initBundle(Bundle bundle) {

        return true;
    }

    protected void initWidget() {

    }

    protected void initPresenter() {

    }

    protected void initData() {

    }


    protected void onRelease() {

    }

    protected void runUiThread(Runnable runnable) {
        runUiThread(runnable, 0);
    }

    protected void runUiThread(Runnable runnable, int delay) {
        View root = this.mRoot;
        if (root == null) return;
        if (Looper.myLooper() == Looper.getMainLooper()) {
            if (delay > 0) {
                root.postDelayed(runnable, delay);
            } else {
                runnable.run();
            }
        } else {
            root.postDelayed(runnable, delay);
        }
    }

    protected void showToast(String message) {
        runOnUiThread(() -> ToastHelper.show(message));
    }

    protected void showToast(@StringRes int messageId) {
        showToast(getString(messageId));
    }

    protected void showCenterToast(String message) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            runOnUiThread(() -> ToastHelper.show(this, message, Gravity.CENTER));
        } else {
            ToastHelper.show(this, message, Gravity.CENTER);
        }
    }

    protected void showCenterToast(@StringRes int messageId) {
        showCenterToast(getString(messageId));
    }

}

File: common/src/main/java/com/sumian/common/container/base/BaseActivity.kt
package com.sumian.common.container.base

import android.content.Intent
import android.content.pm.ActivityInfo
import android.os.Build
import android.os.Bundle
import android.os.Looper
import android.os.MessageQueue
import androidx.appcompat.app.AppCompatActivity
import com.sumian.common.container.IContainer
import com.sumian.common.container.delegate.ContainerDelegate
import com.sumian.common.container.delegate.IContainerDelegate
import com.sumian.common.container.mvp.BaseShowLoadingView
import com.sumian.common.container.mvp.IPresenter

/**
 * Created by sm
 *
 * on 2018/12/6
 *
 * desc:  不可导航的 activity container
 *
 */
abstract class BaseActivity<Presenter : IPresenter> : AppCompatActivity(), BaseShowLoadingView, IContainer {

    companion object {
        private val TAG = BaseActivity::class.java.simpleName
    }

    protected var mPresenter: Presenter? = null

    private val containerDelegate: IContainerDelegate by lazy {
        return@lazy ContainerDelegate.install()
    }

    private val mPrepareInitDataTask by lazy {
        MessageQueue.IdleHandler {
            initData()
            return@IdleHandler false
        }
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        this.intent = intent
        checkBundle(this.intent)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        containerDelegate.setup(this@BaseActivity)
        checkBundle(intent)
        setContentView(getLayoutId())
        initWidgetBefore()
        initWidget()
        prepareDataTask()
    }

    override fun onStop() {
        onCancel()
        super.onStop()
    }

    override fun onDestroy() {
        onRelease()
        super.onDestroy()
    }

    override fun initWidgetBefore() {
    }

    override fun initBundle(bundle: Bundle) {
    }

    override fun initWidget() {
    }

    override fun initData() {
    }

    override fun onRelease() {
        mPresenter?.onRelease()
        containerDelegate.onRelease()
    }

    override fun onCancel() {
        mPresenter?.onCancel()
        containerDelegate.onCancel()
    }

    override fun onShowLoading() {
        containerDelegate.onShowLoading()
    }

    override fun onDismissLoading() {
        containerDelegate.onDismissLoading()
    }

    private fun prepareDataTask() {
        prepareTask(mPrepareInitDataTask)
    }

    private fun checkBundle(intent: Intent?) {
        intent?.extras?.let {
            initBundle(it)
        }
    }

    private fun prepareTask(task: MessageQueue.IdleHandler) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            mainLooper.queue.addIdleHandler(task)
        } else {
            Looper.myQueue().addIdleHandler(task)
        }
    }

    private fun prepareTask(block: () -> Unit) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            mainLooper.queue.addIdleHandler {
                block.invoke()
                false
            }
        } else {
            Looper.myQueue().addIdleHandler {
                block.invoke()
                false
            }
        }
    }

}